File: src/stack_queue/Solution155.java
package stack_queue;

// 155. Min Stack

import java.util.*;

class MinStack {

    Deque<Integer> stack;
    Deque<Integer> min_stack;

    public MinStack() {
        this.stack = new LinkedList<>();
        this.min_stack = new LinkedList<>();
    }

    public void push(int val) {
        if (min_stack.isEmpty() || val <= min_stack.peek()) {
            min_stack.push(val);
        }
        stack.push(val);
    }

    public void pop() {
        if (Objects.equals(min_stack.peek(), stack.pop())) {
            min_stack.pop();
        }
    }

    public int top() {
        return stack.peek();
    }

    public int getMin() {
        return min_stack.peek();
    }
}


public class Solution155 {
}

File: python/stack_queue/Solution155.py
# 155. 最小栈

class MinStack:

    def __init__(self):
        self.stack = []
        self.min_stack = []

    def push(self, val: int) -> None:
        if (not self.min_stack) or self.min_stack[-1] >= val:
            self.min_stack.append(val)
        self.stack.append(val)

    def pop(self) -> None:
        if self.min_stack[-1] == self.stack.pop():
            self.min_stack.pop()

    def top(self) -> int:
        if self.stack:
            return self.stack[-1]
        else:
            return None

    def getMin(self) -> int:
        if self.min_stack:
            return self.min_stack[-1]
        else:
            return None
